src/ast/semantic_analyser.cpp

```diff
@@ -53,6 +53,7 @@ void SemanticAnalyser::warning(const std::string &msg, const location &loc)
 void SemanticAnalyser::visit(Integer &integer)
 {
   integer.type = SizedType(Type::integer, 8, true);
+  integer.type.addrspace = AddrSpace::kernel;
```

Comment:
Yes, I don't think that turning an int into a pointer makes sense as the pointee type information is missing. I got a patch somewhere that makes that disallows it, want to put that up as RFC at some point :)

Reply:
I see, and it is reasonable. So, the problem of

```
sudo  ./src/bpftrace -b -e 'uprobe:/bin/bash:readline { printf("PS1: %s\n", str(*(uptr(curtask->mm->mmap->vm_start) + uaddr("ps1_prompt")))); }'
```

is there is no information about uaddr returning an address of a `char*` value. I feel

```
str(*(int8*)(uptr(curtask->mm->mmap->vm_start) + uaddr("ps1_prompt"))
```

a bit redundant, but can accept it.

Reply:
ah, no the above should be
```
str(uptr(*(uptr(curtask->mm->mmap->vm_start) + uaddr("ps1_prompt")))
```

Reply:
I think `str(curtask->mm->mmap->vm_start + uaddr("ps1_prompt"))` should work, depending on how you define it. In C adding to a pointer is array like:

```
int main() {
  int * a = &G;
  char * b = (char*)&G;
  printf("%p %p\n", a, a+5);
  printf("%p %p\n", b, b+5);
}
----
0x10aeef010 0x10aeef024
0x10aeef010 0x10aeef015
```


`uptr(*A)` seems weird as you're dereferencing it just to turn it back into a pointer again. In this PR `uptr()` should only work on pointers.

Reply:
> uptr(*A) seems weird

yes, but "ps1_prompt" is `char *`, and `uaddr()` returns address, so the type of `uaddr("ps1_prompt")` is `char **`. Therefore `str(*A)` gives the string. The addrspace of `*A` is none (since bpftrace think it's integer), so we need to use uptr to convert it to userspace pointer and It ended up being `str(uptr(*A))`. (I confused at first..) 
bpftrace should treat `*uaddr("ps1_prompt")` as `char *` in this case, but I guess there is no easy way to get the actual type of a symbol.

Reply:
oh yes fair point. But in that case it think it is beter to have `uaddr("ps1_prompt", char *)` or `(char **)uaddr()` as options instead of doing the pointer mess at the end. 
But doing so would mean having a real pointer type in the language (one that stores the type of the pointee, not its size), which is something im working on :)